backend/src/main/java/com/rota/api/StaffService.java

```diff
@@ -95,6 +95,32 @@
         .collect(Collectors.toList());
   }
 
+  private Optional<Engagement> fromDto(EngagementDto engagement) {
+    Optional<Staff> staffOptional = staffRepository.findById(engagement.getStaffId());
+    return staffOptional.map((Staff staff) -> 
```

Comment:
What's the difference of using map vs `ifPresentOrElse`? If the staff member isn't found for example what happens here?

Reply:
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-

If the optional is empty, it will return empty optional, otherwise it will apply the function to the value.
Optional is java's attempt of introducing elements of functional programming and it's an example of a functor (a wrapper object with a map operation defined on it).

I used `map` here, because I want an Optional back `ifPresent` is a `void` method. They are analogous to `map` and `forEach` in javascript arrays.

Reply:
Okay nice, so in this case if `staffOptional` is empty then an empty `Engagement` optional is returned. But if it is present, then it will perform the mapping and the Engagement builder returns an engagement optional? Or does it return an Engagement which gets turned into an optional automatically?

Reply:
`Optional(x).map(f) = Optional(f(x))`
i.e. it takes the contents of the Optional, applies the function (which in this case builds Engagement) and wraps it in an Optional.